hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/FlushTableProcedure.java

```diff
@@ -0,0 +1,172 @@
+
+import java.io.IOException;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.client.RegionReplicaUtil;
+import org.apache.hadoop.hbase.procedure.flush.MasterFlushTableProcedureManager;
+import org.apache.hadoop.hbase.procedure2.ProcedureStateSerializer;
+import org.apache.hadoop.hbase.procedure2.ProcedureSuspendedException;
+import org.apache.hadoop.hbase.procedure2.ProcedureYieldException;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hbase.thirdparty.com.google.protobuf.UnsafeByteOperations;
+
+import org.apache.hadoop.hbase.shaded.protobuf.ProtobufUtil;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProcedureProtos.FlushTableProcedureStateData;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProcedureProtos.FlushTableState;
+
+@InterfaceAudience.Private
+public class FlushTableProcedure extends AbstractStateMachineTableProcedure<FlushTableState> {
+  private static final Logger LOG = LoggerFactory.getLogger(FlushTableProcedure.class);
+
+  private TableName tableName;
+
+  private byte[] columnFamily;
+
+  public FlushTableProcedure() {
+    super();
+  }
+
+  public FlushTableProcedure(MasterProcedureEnv env, TableName tableName) {
+    this(env, tableName, null);
+  }
+
+  public FlushTableProcedure(MasterProcedureEnv env, TableName tableName, byte[] columnFamily) {
+    super(env);
+    this.tableName = tableName;
+    this.columnFamily = columnFamily;
+  }
+
+  @Override
+  protected LockState acquireLock(MasterProcedureEnv env) {
+    // Here we don't acquire table lock because the flush operation and other operations (like
+    // split or merge) are not mutually exclusive. Region will flush memstore when being closed.
+    // It's safe even if we don't have lock.
+    return LockState.LOCK_ACQUIRED;
+  }
+
+  @Override
+  protected void releaseLock(MasterProcedureEnv env) {
+    // nothing to do since we don't acquire lock
+  }
+
+  @Override
+  protected Flow executeFromState(MasterProcedureEnv env, FlushTableState state)
+    throws ProcedureSuspendedException, ProcedureYieldException, InterruptedException {
+    LOG.info("{} execute state={}", this, state);
+
+    try {
+      switch (state) {
+        case FLUSH_TABLE_PREPARE:
+          preflightChecks(env, true);
+          setNextState(FlushTableState.FLUSH_TABLE_FLUSH_REGIONS);
+          return Flow.HAS_MORE_STATE;
+        case FLUSH_TABLE_FLUSH_REGIONS:
+          addChildProcedure(createFlushRegionProcedures(env));
+          return Flow.NO_MORE_STATE;
+        default:
+          throw new UnsupportedOperationException("unhandled state=" + state);
+      }
+    } catch (IOException e) {
+      setFailure("master-flush-table", e);
+    }
+    return Flow.HAS_MORE_STATE;
+  }
+
+  @Override
+  protected void rollbackState(MasterProcedureEnv env, FlushTableState flushTableState)
```

Comment:
Better throw UnsupportedOperationException if this should not happen.

Reply:
OK.